server/src/main/java/org/opensearch/plugins/IdentityAwarePlugin.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.identity.PluginSubject;
+import org.opensearch.identity.Subject;
+
+/**
+ * Plugin that performs transport actions with a plugin system context. IdentityAwarePlugins are initialized
+ * with a {@link Subject} that they can utilize to perform transport actions outside the default subject.
+ *
+ * When the Security plugin is installed, the default subject is the authenticated user. In particular,
+ * SystemIndexPlugins utilize the {@link Subject} to perform transport actions that interact with system indices.
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public interface IdentityAwarePlugin {
+
+    /**
+     * Passes necessary classes for this plugin to operate as an IdentityAwarePlugin
+     *
+     * @param pluginSubject A subject for running transport actions in the plugin context for system index
+     *                      interaction
+     */
+    default void assignSubject(PluginSubject pluginSubject) {}
```

Comment:
Why does a plugin need to identify itself as identity-aware? Rather than having the plugin accept this change, could we add an api on identity service/ plugin service that would return the identity aware node client?

Reply:
Whatever is provided to the plugin needs to be provided on bootstrap and can't be initiated by the plugin.

In extensions, the service account token is provided to an extension on initial handshake so that it can identify itself. Similarly, assignSubject is done on bootstrap so that the plugin can utilize this object to execute transport actions outside the authenticated user context.

I think this either needs to be supplied to createComponents (breaking change) or a separate interface like this.

Reply:
An argument for a separate interface like this is that not all plugins need this PluginSubject. This is particularly needed for SystemIndexPlugins.

Reply:
> Also, the identity goes way beyond just node client
I don't know that we've got alignment on the problem are we trying to solve with this change.  Sure there are way more ways identity could be used by plugins, but I'd rather not put generic platform components in if we don't have a clear picture how they should be used.

Taking a slight step back from this identity focused PR, it sounds like this change is proposing a replacement for `createComponents` with individual composable elements captured by a new set of interfaces (like this one).  If so then maybe we should make sure this model will work well for all things we'd want to compose in.  What about injected items, or could we move this to injection instead?

Reply:
I don't want it to be lost - I think it is a great idea to have composable elements available for plugin dependencies.

Reply:
@peternied Its not proposing a replacement for createComponents. Its creating a replacement for `try (ThreadContext.StoredContext ctx = threadContext.stashContext()) { ... }`.

The problem this PR is solving (along with companion [Security PR](https://github.com/opensearch-project/security/pull/4665)) is making sure that authz checks are enforced. 

This PR and that PR alone will not solve the problem, but pave the way to a long term fix in future versions of OpenSearch.

Reply:
FYI The plan for this to not get lost is to enforce that `ThreadContext.stashContext()` is not used.

In a previous PR @reta and I put other ThreadContext methods behind a JSM permission like this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L224-L232

I plan to do the same to `stashContext()` and enforce is in 3.0.0. Plugins would be able to use it, but only by explicitly granting access to do so in the `plugin-security.policy` file.

By using the `pluginSubject.runAs(() -> { ...your code her... })`, the plugin would not need to add a grant in the `plugin-security.policy` file.

Reply:
> I don't want it to be lost - I think it is a great idea to have composable elements available for plugin dependencies.

I think this is not the goal (at least for this change and I don't see any attempts towards it) but to make an identity first class citizen which is backed by the plugin.

Reply:
Thanks for the thoughts, this sounds like a decent place to be for this PR.  Resolving this thread